src/mscorlib/src/System/Collections/Generic/List.cs

```diff
@@ -1031,6 +1028,45 @@ public void Sort(IComparer<T> comparer)
             return true;
         }
 
+        private void AddEnumerable(IEnumerable<T> enumerable)
+        {
+            Contract.Assert(enumerable != null);
+            Contract.Assert(!(enumerable is ICollection<T>), "We should have optimized for this beforehand.");
+
+            using (IEnumerator<T> en = enumerable.GetEnumerator())
+            {
+                T[] items = _items;
+                int size = _size;
+
+                // Read in items from the enumerator, only updating fields
+                // when we run out of space.
+
+                try
+                {
+                    while (en.MoveNext())
+                    {
+                        if (size == items.Length)
+                        {
+                            _size = size;
+                            EnsureCapacity(size + 1);
+                            items = _items;
+                        }
+
+                        // Note: It's important we increment size after Current is called.
+                        // If that throws an exception we don't want to do the increment.
+                        items[size] = en.Current;
+                        size++;
+                    }
+                }
+                finally
+                {
+                    // Make a final update to _size and _version after we've finished adding.
+                    _size = size;
+                    _version++;
```

Comment:
> However if you have done manual inlining enough times, the real world workloads would likely get slower because of the code size increase from manual inlining would overwhelm the benefits.

I think it makes sense in this case. This in a loop and it's part of the bottleneck. It's also frequently used (directly, or indirectly by Linq).

> We may want to be more data driven for manual inlining - only do it if there is real-world workload where it makes a dent,

It just came up in corefx in a frequently-used method. See the PR I linked to: https://github.com/dotnet/corefx/pull/13942#issuecomment-262844479